src/pkg/packager/publish.go

```diff
@@ -0,0 +1,337 @@
+	if err != nil {
+		return err
+	}
+	componentTarballs := []string{}
+
+	// repackage the component directories into tarballs
+	for _, componentDir := range componentDirs {
+		dst := filepath.Join(p.tmp.Components, filepath.Base(componentDir)+".tar")
+		err = archiver.Archive([]string{componentDir}, dst)
+		if err != nil {
+			return err
+		}
+		componentTarballs = append(componentTarballs, dst)
+		_ = os.RemoveAll(componentDir)
+	}
+	paths = append(paths, componentTarballs...)
+	imagesLayers, err := filepath.Glob(filepath.Join(p.tmp.Images, "blobs", "sha256", "*"))
+	if err != nil {
+		return err
+	}
+	paths = append(paths, imagesLayers...)
+	ref, err := p.ref("")
+	if err != nil {
+		return err
+	}
+	message.HeaderInfof("📦 PACKAGE PUBLISH %s:%s", p.cfg.Pkg.Metadata.Name, ref.Reference)
+	err = p.publish(ref, paths)
+	if err != nil {
+		return fmt.Errorf("unable to publish package %s: %w", ref, err)
+	}
+	return nil
+}
+
+func (p *Packager) publish(ref registry.Reference, paths []string) error {
+	message.Infof("Publishing package to %s", ref)
+	spinner := message.NewProgressSpinner("")
+	defer spinner.Stop()
+
+	// destination remote
+	dst, err := utils.NewOrasRemote(ref)
+	if err != nil {
+		return err
+	}
+	ctx := dst.Context
+
+	// source file store
+	src, err := file.New(p.tmp.Base)
+	if err != nil {
+		return err
+	}
+	defer src.Close()
+
+	var descs []ocispec.Descriptor
+
+	for idx, path := range paths {
+		name, err := filepath.Rel(p.tmp.Base, path)
+		if err != nil {
+			return err
+		}
+		spinner.Updatef("Preparing layer %d/%d: %s", idx+1, len(paths), name)
+
+		mediaType := parseZarfLayerMediaType(name)
+
+		desc, err := src.Add(ctx, name, mediaType, path)
+		if err != nil {
+			return err
+		}
+		descs = append(descs, desc)
+	}
+	spinner.Successf("Prepared %d layers", len(descs))
+
+	copyOpts := oras.DefaultCopyOptions
+	copyOpts.Concurrency = p.cfg.PublishOpts.CopyOptions.Concurrency
+	copyOpts.OnCopySkipped = utils.PrintLayerExists
+	copyOpts.PostCopy = utils.PrintLayerExists
+
+	var root ocispec.Descriptor
+
+	// try to push an ArtifactManifest first
+	// not every registry supports ArtifactManifests, so fallback to an ImageManifest if the push fails
+	// see https://oras.land/implementors/#registries-supporting-oci-artifacts
+	root, err = p.publishArtifact(dst, src, descs, copyOpts)
+	if err != nil {
+		// reset the progress bar between attempts
+		dst.ProgressBar.Stop()
+
+		// log the error, the expected error is a 400 manifest invalid
+		message.Debug("ArtifactManifest push failed with the following error, falling back to an ImageManifest push:", err)
+
+		// if the error returned from the push is not an expected error, then return the error
+		if !isManifestUnsupported(err) {
+			return err
+		}
+		// fallback to an ImageManifest push
+		root, err = p.publishImage(dst, src, descs, copyOpts)
+		if err != nil {
+			return err
+		}
+	}
+	dst.ProgressBar.Successf("Published %s [%s]", ref, root.MediaType)
```

Comment:
There is an error here with how the progressbar renders (non-blocking but another issue should be created if we decide not to tackle)
![image](https://user-images.githubusercontent.com/3977569/225154301-4452ff56-062f-44d3-97ad-1b4748756390.png)